driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -483,24 +488,62 @@ void setKeyspace(String keyspace) throws ConnectionException {
         }
     }
 
-    ListenableFuture<Void> setKeyspaceAsync(final String keyspace) throws ConnectionException, BusyConnectionException {
-        logger.trace("{} Setting keyspace {}", this, keyspace);
-        // Note: we quote the keyspace below, because the name is the one coming from Cassandra, so it's in the right case already
-        Future future = write(new Requests.Query("USE \"" + keyspace + '"'));
-        return Futures.transform(future, new AsyncFunction<Message.Response, Void>() {
-            @Override
-            public ListenableFuture<Void> apply(Message.Response response) throws Exception {
-                if (response instanceof SetKeyspace) {
-                    Connection.this.keyspace = ((SetKeyspace) response).keyspace;
-                    return MoreFutures.VOID_SUCCESS;
-                } else if (response.type == ERROR) {
-                    Responses.Error error = (Responses.Error) response;
-                    throw defunct(error.asException(address));
-                } else {
-                    throw defunct(new DriverInternalError("Unexpected response while setting keyspace: " + response));
+    ListenableFuture<Connection> setKeyspaceAsync(final String keyspace) throws ConnectionException, BusyConnectionException {
+        if (Objects.equal(keyspace(), keyspace)) {
+            return thisFuture;
+        }
```

Comment:
I made this change, but didn't remove the case from the while loop.  Justification for that is otherwise we'd have to create a new `SetKeyspaceAttempt` every time, whether or not we need to.  Instead a `SetKeyspaceAttempt` is only ever created if the current `targetKeyspace` doesn't match.